cluster-controller/src/main/java/io/strimzi/controller/cluster/operations/KafkaClusterOperation.java

```diff
@@ -0,0 +1,263 @@
+        vertx.sharedData().getLockWithTimeout(lockName, LOCK_TIMEOUT, res -> {
+            if (res.succeeded()) {
+                Lock lock = res.result();
+
+                ClusterDiffResult diff;
+                KafkaCluster kafka;
+                ConfigMap kafkaConfigMap = k8s.getConfigmap(namespace, name);
+
+                if (kafkaConfigMap != null)    {
+
+                    try {
+
+                        kafka = KafkaCluster.fromConfigMap(kafkaConfigMap);
+                        log.info("Updating Kafka cluster {} in namespace {}", kafka.getName(), namespace);
+                        diff = kafka.diff(k8s, namespace);
+
+                    } catch (Exception ex) {
+
+                        log.error("Error while parsing cluster ConfigMap", ex);
+                        handler.handle(Future.failedFuture("ConfigMap parsing error"));
+                        lock.release();
+                        return;
+                    }
+
+                } else {
+                    log.error("ConfigMap {} doesn't exist anymore in namespace {}", name, namespace);
+                    handler.handle(Future.failedFuture("ConfigMap doesn't exist anymore"));
+                    lock.release();
+                    return;
+                }
+
+                Future<Void> chainFuture = Future.future();
+
+                scaleDown(kafka, namespace, diff)
+                        .compose(i -> patchService(kafka, namespace, diff))
+                        .compose(i -> patchHeadlessService(kafka, namespace, diff))
+                        .compose(i -> patchStatefulSet(kafka, namespace, diff))
+                        .compose(i -> patchMetricsConfigMap(kafka, namespace, diff))
+                        .compose(i -> rollingUpdate(kafka, namespace, diff))
+                        .compose(i -> scaleUp(kafka, namespace, diff))
+                        .compose(chainFuture::complete, chainFuture);
+
+                chainFuture.setHandler(ar -> {
+                    if (ar.succeeded()) {
+                        log.info("Kafka cluster {} successfully updated in namespace {}", kafka.getName(), namespace);
+                        handler.handle(Future.succeededFuture());
+                        lock.release();
+                    } else {
+                        log.error("Kafka cluster {} failed to update in namespace {}", kafka.getName(), namespace);
+                        handler.handle(Future.failedFuture("Failed to update Zookeeper cluster"));
+                        lock.release();
+                    }
+                });
+            } else {
+                log.error("Failed to acquire lock to create Kafka cluster {}", lockName);
+                handler.handle(Future.failedFuture("Failed to acquire lock to create Kafka cluster"));
+            }
+        });
+    }
+
+    private Future<Void> scaleDown(KafkaCluster kafka, String namespace, ClusterDiffResult diff) {
```

Comment:
There is a lot of code duplication between Kafka and Zookeeper. Can we do something about it?

Reply:
Probably, again I will have a look at this.